cli/printer.js

```diff
@@ -1,46 +1,183 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 'use strict';
 
-const Printer = {
+const fs = require('fs');
+const Report = require('../report/report');
+
+class Printer {
+
+  /**
+   * @constructor
+   */
+  constructor() {
+    this._outputMode = 'pretty';
+    this._outputPath = 'stdout';
+  }
+
+  /**
+   * An enumeration of acceptable output modes:
+   * <ul>
+   *   <li>'pretty': Pretty print the results</li>
+   *   <li>'json': JSON formatted results</li>
+   *   <li>'html': An HTML report</li>
+   * </ul>
+   *
+   * @return results
+   */
+  static get _outputModes() {
+    return [
+      'pretty',
+      'json',
+      'html'
+    ];
+  }
+
+  /**
+   * @return {string} The current output mode.
+   */
+  get outputMode() {
+    return this._outputMode;
+  }
+
+  /**
+   * The output mode to use.
+   *
+   * @see Printer._outputModes;
+   * @param {string} mode
+   */
+  set outputMode(mode) {
+    if (Printer._outputModes.indexOf(mode) === -1) {
+      console.warn(`Unknown output mode ${mode}; using pretty`);
+      this._outputMode = 'pretty';
+      return;
+    }
+
+    this._outputMode = mode;
+  }
+
+  /**
+   * @return {string} The current output path.
+   */
+  get outputPath() {
+    return this._outputPath;
+  }
+
+  /**
+   * The output path to use, either stdout or a file path.
+   * @param {string} path
+   */
+  set outputPath(path) {
+    if (!path) {
+      console.warn('No output path set; using stdout');
+      this._outputPath = 'stdout';
+      return;
+    }
+
+    this._outputPath = path;
+  }
+
   /**
-   * @param {{info: function(...)}} debugLog
-   * @param {{log: function(...)}} output
-   * @param {string} url
-   * @param {!Array<?>} results
+   * Writes the results.
+   *
+   * @param {{url: string, aggregations: !Array<*>}} results
+   * @return {Promise}
    */
-  json: function(debugLog, output, url, results) {
-    debugLog.info('\n\n\nLighthouse results (JSON):', url);
+  write(results) {
```

Comment:
I feel like the interface for this class is more complicated than needed. What about just `write(mode, path, results)` and have it check the mode and path validity internal to write, then pass the results to the appropriate reporter?